applications/FluidDynamicsApplication/custom_processes/calulate_levelset_consistent_nodal_gradient_process.cpp

```diff
@@ -0,0 +1,148 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Mohammad R. Hashemi
+//
+
+// Project includes
+#include "includes/define.h"
+#include "utilities/parallel_utilities.h"
+
+// Application includes
+#include "fluid_dynamics_application_variables.h"
+#include "calulate_levelset_consistent_nodal_gradient_process.h"
+
+
+namespace Kratos
+{
+
+/* Public functions *******************************************************/
+
+/// constructor
+CalulateLevelsetConsistentNodalGradientProcess::CalulateLevelsetConsistentNodalGradientProcess(
+        ModelPart& rModelPart)
+    : Process(), mrModelPart(rModelPart) {}
+
+/// Constructor with Kratos parameters.
+CalulateLevelsetConsistentNodalGradientProcess::CalulateLevelsetConsistentNodalGradientProcess(
+    ModelPart& rModelPart,
+    Parameters Parameters)
+    : CalulateLevelsetConsistentNodalGradientProcess(
+        rModelPart
+    ){}
+
+/// Constructor with Kratos model
+CalulateLevelsetConsistentNodalGradientProcess::CalulateLevelsetConsistentNodalGradientProcess(
+    Model& rModel,
+    Parameters Parameters)
+    : CalulateLevelsetConsistentNodalGradientProcess(
+        rModel.GetModelPart(Parameters["model_part_name"].GetString())
+    ){}
+
+void CalulateLevelsetConsistentNodalGradientProcess::Execute(){
+
+    KRATOS_TRY;
+
+    const auto zero_vector = ZeroVector(3);
+    // Set to zero
+    block_for_each(mrModelPart.Nodes(), [&](Node<3>& rNode){
+        rNode.SetValue(NODAL_AREA, 0.0);
+        rNode.SetValue(PRESSURE_GRADIENT, zero_vector);
+    });
+
+    // Current domain size
+    const unsigned int num_dim = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
+    const unsigned int num_nodes = num_dim + 1; //For tetrahedra and triangles
+
+    // Auxiliar containers
+    double detJ0 = 0.0;
+    Matrix DN_DX, J0, InvJ0;
+    Vector N(num_nodes), pressures(num_nodes), distances(num_nodes), grad(num_dim);
+
+    // First element iterator
+    const auto it_element_begin = mrModelPart.ElementsBegin();
+
+    const int elements_number = mrModelPart.Elements().size();
+
+    // Iterate over the elements
+    #pragma omp parallel for firstprivate(DN_DX, N, J0, InvJ0, detJ0, pressures, distances, grad)
```

Comment:
The parallel utils now support TLS (thread local storage).

Reply:
The loop **on elements** does multiple modifications to the historical/non-historical variables saved on **nodes**. I think it is still out of the scope of parallel utilities?

Reply:
No. The TLS allows defining a custom local storage container. In this case the custom container involves different types so will require creating a tuplle.

Reply:
Could you please mention any example in the code with similar implementation?

Reply:
I did the same for the compressible NS shock capturing. However it is not merged yet. Once everything is ready let me know and I can change this :wink:

Reply:
I'm compiling the changes in my machine. I'll push the changes as soon as I run the tests.